### ISSUE TITLE
Downloadable Reports

### ISSUE THREAD

**Issue by srerickson** [psu-libraries/scholarsphere#1210]
# Feature Request
We should make it easier for admins and users to download reports (`csv` files) about works, downloads, views, etc. This issue describes the reports that should be accessible through the UI.

- Admin reports
  - [x] Monthly Report 
  - [x] All Works
  - [x] All Work Versions
  - [x] All Files
- User Reports
  - [x] Monthly Report 

## Admin reports

These reports should be accessible through a "reports" admin-only page (linked in top-right menu)

### Monthly Reports

These monthly reports are generated each month. Each month's report is individually downloadable from the reports page. We should generate reports for all previous months (as available data permits).  Report includes one work-month per row, with the following columns:

column | description
-- | --
work_id | UUID of the work
month | numeric representation of month (1-12)
year | year (2022)
downloads | aggregated number of downloads for the work in the month
views | aggregated number of views for the work in the month

### All works

Report includes all Works (one per row), with the following columns:

column | description
-- | --
id | work's UUID
depositor | PSU ID
work_type | example: "Artilce"
title | title from most recent version
doi | scholarsphere's doi
deposited_at | Date draft created
deposit_agreed_at | Date of most recent published version
embargoed_until | date
visibility | "Public" or "Penn State Only"
latest_published_version | UUID of published work version (if present)
downloads | aggregated downloads across all versions/files
views | aggregated views across all versions

### All Work Versions
Report includes all Work Versions (one per row), with the following columns:

column | description
-- | --
id | version's UUID
work_id | parent work's UUID
state | published, draft, withdrawn
version_number | number of the version in the work
title | meta field (joined if list)
subtitle | ..
version_name | ..
keyword | ..
rights | ..
description | ..
publisher_statement | ..
resource_type | ..
contributor | ..
publisher | ..
published_date | ..
subject | ..
language | ..
identifier | ..
based_near | ..
related_url | ..
source | ..
~downloads~ | ~aggregated downloads across all files~
views | number of views

### All Files
Report includes all *File Memberships* (join of file resource and work version, one per row), with the following columns:

column | description
-- | --
id | file resource's UUID
version_id | work version's UUID
filename | name of the file
mime_type | MIME type
size | number of bytes
md5 | md5 checksum (once implemented)
sha256 | sha256 checksum  (once   implemented)
downloads | number of downloads

## User Reports

User reports are scoped to the works that users have deposited. They are linked from the user's profile page.

### Monthly Reports

Report includes one work-month per row, with the following columns:

column | description
-- | --
work_id | UUID of the work
title | title of the work
month | numeric representation of month (1-12)
year | year (2022)
downloads | aggregated number of downloads for the work in the month
views | aggregated number of views for the work in the month

**Comment by srerickson** [comment]
@nicktiberi about how large (in rows) would the "monthly report" (admin version) above be? Too large?

**Comment by nicktiberi** [comment]
> @nicktiberi about how large (in rows) would the "monthly report" (admin version) above be? Too large?

@srerickson - without building the full query to see the actual results, we have about 11k works in the production DB right now, so we would at most get 11k rows per month in the result set, assuming every work was viewed at least once in that month. So the max theoretical rows for a year of results would be ~132k.

The view statistics are currently stored at the WorkVersion level and the FileResource level, not at the Work level, so they would have to be aggregated at the time of the report with some joins. I _think_ that would be doable without being too slow, but would have to refresh my SQL JOIN memory a little bit.

**Comment by rschenk** [comment]
@srerickson in the multi-fields, you say to join them if multiple. How would you like those delimited?

**Comment by srerickson** [comment]
I was imagining some csv embedded as a value in the csv (i.e., double escaped). Hopefully there's not too much of this nonsense.

**Comment by rschenk** [comment]
@srerickson Another thing to note is, in the All Work Versions report, the downloads column may create some confusion. Downloads are tracked on the actual binary file itself (`FileResource` in Scholarsphere lingo) which is shared between many versions of the same work. The number of downloads that would be reported would be the total  number of downloads of that file, summed across all versions. We do not have the means to track downloads of a file within a specific ~work~ version (edit: typo).

In other words, if you had a work with two versions, Version A and Version B, and they both referenced the same file F. Lets say F was downloaded 2x from Version A and 3x  from Version B. In the report, the row representing A and the row representing B would both list 5 as the number of file downloads

**Comment by srerickson** [comment]
@rschenk - that makes sense. Let's just aggregate downloads at the work level then. Thanks for clarifying.

**Comment by nicktiberi** [comment]
**Note:** we talked in standup today and @srerickson is ok with the array values being represented in the CSV as `["value1", "value2"]` etc. so no need to do any special processing for those values at this time.

**Comment by srerickson** [comment]
on the qa server, I'm getting a server error when I download the "all works" report.
